### DISCUSSION TITLE
Universal Build Directory Setting Should be Overridable Per Configuration

### DISCUSSION THREAD

**Discussion by daeden** [microsoft/vscode-cmake-tools#2208] [Ideas]
Currently, changing the build directory requires making a change that impacts all configurations. This setting should have a default value, but it should be overridable per configuration.

![image](https://user-images.githubusercontent.com/2382269/138787952-8660f0b4-be58-4cc2-afd1-215973d94d3f.png)

It is fine to support a default build directory. In fact that's a good idea, but why no support for overriding the default path in the cmake_configurations.json file with a simple setting like:

`"buildDirectory": "${sourceFolder}/my_custom_build_dir"`

It is very frustrating how clumsy it is to switch between configurations. I have to think twice before changing from one config to the other knowing that it is going to eat up a huge amount of time re building from scratch.

As such, I have started using annoying workarounds:

- Manually moving the build directory before changing configs (God help you if you forget)
- Multiple workspaces (very annoying since I have to synchronize the workspaces now)
- Manually change the global build directory before switching configs (also sucks if you forget)

All of which would be mercifully unnecessary if the config file allowed for this setting to be overriden.

**Comment by andreeis** [comment]
@daeden, try using the variable "${buildType}" in the definition of your "buildDirectory". It should expand to the value of the current variant ("debug", "release", ...etc...). Let us know if this helps.

**Reply by daeden** [reply]
brilliant! That did the trick. It would still be great if this particular configuration setting could be specified in the same place where I would configure any other configuration setting, but I can work with this!

Thx.

**Reply by bobbrow** [reply]
> It would still be great if this particular configuration setting could be specified in the same place where I would configure any other configuration setting

What do you mean by this?  You can set this in `.vscode/settings.json` for every folder you open.

**Reply by daeden** [reply]
@bobbrow I'm not sure what you mean by opening multiple folders. I'm coming from CLion, so I'm still getting my head around how VS Code works.

My code is in a directory:

/home/gperez/git/cts

I open that directory and there I have a .vscode folder with all the appropriate json files.

What other directory would I be opening?

**Reply by daeden** [reply]
> @daeden, try using the variable "${buildType}" in the definition of your "buildDirectory". It should expand to the value of the current variant ("debug", "release", ...etc...). Let us know if this helps.

@andreeis 

Unfortunately, this does not work all the way.

In my launch.json I have the following:

` "program": "${config:cmake.buildDirectory}/bin/Series_v2_bars_ut",`

But this expands to:

/home/gperez/git/cts_vscode/build_${buildType}/bin/Series_v2_bars_ut

Any other ideas?

**Reply by andreeis** [reply]
@daeden, in launch.json we recommend using variables like ${command:cmake.launchTargetPath} or getLaunchTargetPath (which will not trigger a build before debugging or launching in terminal). That variable will point precisely to the current debug target that is selected and you would not need to deduce it using formulas with ${buildType} for "program" launch.json field.

Let us know how that works.

**Reply by daeden** [reply]
@andreeis thanks!

Everything is working nicely!
